test/jdk/sun/security/tools/keytool/GenKeyPairSigner.java

```diff
@@ -0,0 +1,319 @@
+                .shouldHaveExitValue(0);
+
+        // examine the resulting cert
+        kstore = KeyStore.getInstance(new File("ks"), "changeit".toCharArray());
+        cert = (X509Certificate)kstore.getCertificate("e2");
+        sigName = cert.getSigAlgName();
+        if (sigName != "SHA384withRSA") {
+            System.out.println("Signature algorithm name is in error");
+            System.exit(1);
+        }
+
+        pKey = cert.getPublicKey();
+        keyLen = KeyUtil.getKeySize(pKey);
+        if (keyLen != 448) {
+            System.out.println("Key size is in error");
+            System.exit(1);
+        }
+
+        pKeyAlg = pKey.getAlgorithm();
+        if (pKeyAlg != "XDH") {
+            System.out.println("Subject Public Key Algorithm is in error");
+            System.exit(1);
+        }
+
+        SecurityTools.keytool("-keystore ks -storepass changeit " +
+                "-list -v")
+                .shouldContain("Alias name: e2")
+                .shouldContain("Signature algorithm name: SHA384withRSA")
+                .shouldContain("Subject Public Key Algorithm: 448-bit XDH key")
+                .shouldHaveExitValue(0);
+
+        kt("-genkeypair -keyalg DSA -alias ca3 -dname CN=CA3 -ext bc:c ",
+                "ks");
+
+        System.out.println("Generating a DH cert with -signer option");
+        SecurityTools.keytool("-keystore ks -storepass changeit " +
+                "-genkeypair -keyalg DH -alias e3 -dname CN=E3 -signer ca3")
+                .shouldContain("Generating 2,048 bit DH key pair and a certificate (SHA256withDSA) issued by an entry specified by the -signer option with a validity of 90 days")
+                .shouldContain("for: CN=E3")
+                .shouldHaveExitValue(0);
+
+        // examine the resulting cert
+        kstore = KeyStore.getInstance(new File("ks"), "changeit".toCharArray());
+        cert = (X509Certificate)kstore.getCertificate("e3");
+        sigName = cert.getSigAlgName();
+        if (sigName != "SHA256withDSA") {
+            System.out.println("Signature algorithm name is in error");
+            System.exit(1);
+        }
+
+        pKey = cert.getPublicKey();
+        keyLen = KeyUtil.getKeySize(pKey);
+        if (keyLen != 2048) {
+            System.out.println("Key size is in error");
+            System.exit(1);
+        }
+
+        pKeyAlg = pKey.getAlgorithm();
+        if (pKeyAlg != "DH") {
+            System.out.println("Subject Public Key Algorithm is in error");
+            System.exit(1);
+        }
+
+        SecurityTools.keytool("-keystore ks -storepass changeit " +
+                "-list -v")
+                .shouldContain("Alias name: e3")
+                .shouldContain("Signature algorithm name: SHA256withRSA")
+                .shouldContain("Subject Public Key Algorithm: 2048-bit DH key")
+                .shouldHaveExitValue(0);
+    }
+
+    static void testSignerJKS() throws Exception {
```

Comment:
Please add some comments on why testing with JKS is necessary.

Reply:
Added the comments about JKS keystore.